pallets/pips/src/lib.rs

```diff
@@ -809,194 +864,264 @@ decl_module! {
 
-            // Close proposal
-            Self::update_referendum_state(id, ReferendumState::Rejected);
-            Self::prune_data(id, Self::prune_historical_pips());
+            // 5. Emit event.
+            Self::deposit_event(RawEvent::SnapshotTaken(did));
             Ok(())
         }
 
-        /// It updates the enactment period of a specific referendum.
+        /// Enacts `results` for the PIPs in the snapshot queue.
+        /// The snapshot will be available for further enactments until it is cleared.
         ///
-        /// # Arguments
-        /// * `until`, It defines the future block where the enactment period will finished.  A
-        /// `None` value means that enactment period is going to finish in the next block.
+        /// The `results` are encoded a list of `(id, result)` where `result` is applied to `id`.
+        /// Note that the snapshot priority queue is encoded with the *lowest priority first*.
+        /// so `results = [(id, Approve)]` will approve `SnapshotQueue[SnapshotQueue.len() - 1]`.
         ///
         /// # Errors
-        /// * `BadOrigin`, Only the release coordinator can update the enactment period.
-        /// * ``,
-        #[weight = (750_000_000, DispatchClass::Operational, Pays::Yes)]
-        pub fn override_referendum_enactment_period(origin, id: PipId, until: Option<T::BlockNumber>) -> DispatchResult {
-            let sender = ensure_signed(origin)?;
-            let current_did = Context::current_identity_or::<Identity<T>>(&sender)?;
-
-            // 1. Only release coordinator
-            ensure!(
-                Some(current_did) == T::GovernanceCommittee::release_coordinator(),
-                Error::<T>::BadOrigin);
-
-            Self::is_referendum_state(id, ReferendumState::Scheduled)?;
+        /// * `BadOrigin` - unless a GC voting majority executes this function.
+        /// * `CannotSkipPip` - a given PIP has already been skipped too many times.
+        /// * `SnapshotResultTooLarge` - on len(results) > len(snapshot_queue).
+        /// * `SnapshotIdMismatch` - if:
+        ///   ```
+        ///    ∃ (i ∈ 0..SnapshotQueue.len()).
+        ///      results[i].0 ≠ SnapshotQueue[SnapshotQueue.len() - i].id
+        ///   ```
+        ///    This is protects against clearing queue while GC is voting.
+        #[weight = (1_000_000_000, DispatchClass::Operational, Pays::Yes)]
+        pub fn enact_snapshot_results(origin, results: Vec<(PipId, SnapshotResult)>) -> DispatchResult {
+            T::VotingMajorityOrigin::ensure_origin(origin)?;
 
-            // 2. New value should be valid block number.
-            let next_block = <system::Module<T>>::block_number() + 1.into();
-            let new_until = until.unwrap_or(next_block);
-            ensure!( new_until >= next_block, Error::<T>::InvalidFutureBlockNumber);
+            let max_pip_skip_count = Self::max_pip_skip_count();
+
+            <SnapshotQueue<T>>::try_mutate(|queue| {
```

Comment:
Does this fail cleanly if there is no SnapshotQueue?

Reply:
Per discussion on slack: `enact_shapshot_results(empty_vec);` should always be a noop, but we want a test to ensure this.

Reply:
As per Slack discussion, the answer is yes... just needs a test case to prove it ;-)